### PR TITLE
add gauge metric for DAG cache items

### LINKED ISSUE #5005: add metrics for DAG cache contents
It'd be nice to have ~~counter~~ gauge metrics indicating how many items are in the DAG cache (by kind at least). This would be helpful for ensuring we aren't holding onto resources we shouldn't be.

Suggested structure would be a single GaugeVec metric, named `contour_dag_cache_object_count`, with at least one label, `kind`, and potentially another for `namespace`. This metric should be set each time an object is added to or removed from the DAG cache, based on the current length of the relevant map.

### PR DESCRIPTION
**Pull request by izturn**
Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)

fix #5005 

BTW: 
 the prometheus complain with:
 > contour_dag_cache_object_count: non-histogram and non-summary metrics should not have "_count" suffix, 

so remove the suffix: `_count`

### CODE REVIEW

File: changelogs/unreleased/5109-izturn-small.md
Comment (skriss) [inline changelogs/unreleased/5109-izturn-small.md:1]:
```suggestion
Adds a new gauge metric, `contour_dag_cache_object`, to indicate the total number of items that are currently in the DAG cache.
```

File: internal/metrics/metrics.go
Comment (skriss) [inline internal/metrics/metrics.go:37]:
```suggestion
	dagCacheObjectGauge         *prometheus.GaugeVec
```

File: internal/metrics/metrics.go
Comment (skriss) [inline internal/metrics/metrics.go:70]:
```suggestion
	DAGCacheObjectGauge         = "contour_dag_cache_object"
```

File: internal/metrics/metrics.go
Comment (skriss) [inline internal/metrics/metrics.go:237]:
```suggestion
func (m *Metrics) SetDAGCacheObjectMetric(kind string, count int) {
```

File: internal/metrics/metrics_test.go
Comment (skriss) [inline internal/metrics/metrics_test.go:695]:
```suggestion
func TestSetDAGCacheObjectMetric(t *testing.T) {
```

File: internal/metrics/metrics.go
Comment (skriss) [inline internal/metrics/metrics.go:138]:
```suggestion
				Help: "Total number of items that are currently in the DAG cache.",
```

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for the PR @izturn! A few small things to address at first.

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:233]: We need to set this regardless of whether `ok` was true or not, since `ok` represents whether the DAG needs to be rebuilt or not, _not_ whether something was added to the cache or not.

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:269]: same here, need to set the metric whether `ok` was true or not.

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:133]:
```suggestion
				return false, len(kc.ingresses)
```

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:149]:
```suggestion
				return false, len(kc.httpproxies)
```

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:165]: The returned value here should be 1 if `kc.gatewayclass` is not nil

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:182]: The returned value here should be 1 if `kc.gateway` is not nil

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:316]: 1 if kc.gatewayclass is not nil, otherwise 0

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:329]: 1 if `kc.gateway` is not nil, otherwise 0

File: internal/dag/cache.go
Comment (skriss) [inline internal/dag/cache.go:232]: Please double-check that `KindOf` actually covers all the resource types we cache -- IIRC it was missing some (ref. #5012)

**Review by skriss — CHANGES_REQUESTED** [review]

File: internal/dag/cache.go
Comment (izturn) [reply internal/dag/cache.go]: yes, the `KindOf` not covers the all resource types now, but I don't know whether I should update it or leave it to #5012

Comment (skriss) [reply internal/dag/cache.go]: OK let's get #5012 merged and then we'll be good here.

**Comment by skriss** [comment]
@izturn this needs a rebase as we added GRPCRoute support, please make sure we have a metric for that in the DAG cache as well!

**Comment by izturn** [comment]
@skriss PTAL

**Review by skriss — APPROVED** [review]
Tested this out and LGTM, thanks @izturn!

**Comment by skriss** [comment]
Sorry, I should've held my #5133 until this merged, but now a minor merge conflict to resolve (we no longer are informing on ContourConfiguration, so don't need the logic for it in the DAG cache)

File: internal/dag/cache.go
Comment (sunjayBhatia) [inline internal/dag/cache.go:239]: if we keep it with this kind of implementation, what about instead of finding the count of resources in `maybeInsert`, we just take the kind and take the length of the map of the appropriate resource here (or even just iterate over all the kinds and set them all)

finding the number of resources to set the metric to above seems like it's easy to make a small mistake that is hard to catch, esp., in cases with special handling of resources that we don't necessarily want to add etc.

it would be good to add some testing to make sure we don't make any such mistakes

**Review by sunjayBhatia — COMMENTED** [review]
Apologies, I'm a little late on this one but this seems like an opportunity to pull this out into it's own component/abstraction so we can test things and so this functionality doesn't interfere with the dag cache directly and we don't need to change the dag itself to get this metric

e.g. in other places we use the `ResourceEventHandler` or `dag.Observer` interface to compose components together

we could have an `Observer` or other abstraction that runs after the cache is updated to take stock of what is in the cache and send metrics

see https://github.com/projectcontour/contour/blob/main/internal/contour/metrics.go

File: internal/dag/cache.go
Comment (izturn) [reply internal/dag/cache.go:239]: @sunjayBhatia  thx,  @skriss WDYT? I will resend a new pr or update this one after we make a decision

Comment (sunjayBhatia) [reply internal/dag/cache.go:239]: chatting with @skriss offline, I'm revising my position and in support of this impl:
- with my suggestion if any new resource kinds are added to the cache you will have to remember to update the component that sets the cache metrics, this impl forces you to b/c of the signature of `maybeInsert` etc.
- in the future we plan to attempt to utilize go generics more which should give us opportunities to refactor this

Comment (sunjayBhatia) [reply internal/dag/cache.go:239]: when I get some time I'm planning on experimenting with making KubernetesCache use generics to simplify things and reduce type assertions/switches etc.

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**
